### PR TITLE
config: add system.store-certs.[a-zA-Z0-9] support

### PR DESCRIPTION
**Pull request by mvo5**
This option allows adding pem encoded ssl certificates
to the snapd config that will then be used by snapd.

Certificates are added via:
```
$ sudo snap set system store-certs.extra1="$(cat mykey.pem)"
```
and removed via:
```
$ sudo snap unset system 'store-certs.extra1'
```

The certificates are written to /var/lib/snapd/ssl/store-certs
and picked up by the httputil.NewHTTPClient(). This
means that they will be picked up by snap-repair as
well automatically.

Snapd on classic will also honor the ssl config but
it's less useful here because there are distro mechanisms
already to add certificates.

### CODE REVIEW

File: overlord/configstate/configcore/certs_test.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+		},
+		changes: map[string]interface{}{
+			"certs.certthatwillbereverted": mockCert,
+		},
+	})
+	c.Assert(err, IsNil)
+	c.Assert(filepath.Join(dirs.SnapdStoreSSLCertsDir, "cert1.pem"), testutil.FilePresent)
+	c.Assert(filepath.Join(dirs.SnapdStoreSSLCertsDir, "certthatwillbereverted.pem"), testutil.FilePresent)
+
+	// now simulate a "snap revert core" where "cert1" will stay in
```
Comment (anonymouse64) [inline overlord/configstate/configcore/certs_test.go]: nice test!

File: overlord/configstate/configcore/certs.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+		if err != nil {
+			return fmt.Errorf("internal error: cannot get data for %s: %v", nameWithoutSnap, err)
+		}
+		if cert != "" {
+			optionName := strings.SplitN(name, ".", 3)[2]
+			block, rest := pem.Decode([]byte(cert))
+			if block == nil || block.Type != "CERTIFICATE" || len(rest) > 0 {
+				return fmt.Errorf("cannot decode pem certificate %q", optionName)
+			}
+			// XXX: add more validations?
```
Comment (pedronis) [inline overlord/configstate/configcore/certs.go]: yes, we should do the checks that

* len(block.Headers) == 0
* ParseCertificate(block.Bytes)

like AppendCertsFromPEM

File: overlord/configstate/configcore/certs.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+		nameWithoutSnap := strings.SplitN(name, ".", 2)[1]
+		cert, err := coreCfg(tr, nameWithoutSnap)
+		if err != nil {
+			return fmt.Errorf("internal error: cannot get data for %s: %v", nameWithoutSnap, err)
+		}
+		if cert != "" {
+			optionName := strings.SplitN(name, ".", 3)[2]
+			block, rest := pem.Decode([]byte(cert))
+			if block == nil || block.Type != "CERTIFICATE" || len(rest) > 0 {
```
Comment (pedronis) [inline overlord/configstate/configcore/certs.go]: the certs could be a chain I think in which case we need to check rest and not disallow it

File: overlord/configstate/configcore/certs.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/snapcore/snapd/overlord/configstate/config"
+)
+
+func handleCertConfiguration(tr config.Conf) error {
+	// This handles the "snap revert core" case:
+	// We need to go over each pem cert on disk and check if there is
+	// a matching config entry - if not->delete the cert
+	//
+	// XXX: remove this code once we have a general way to handle
+	//      "snap revert" and config updates
```
Comment (pedronis) [inline overlord/configstate/configcore/certs.go]: even before we consider that, there's probably something cheap we can do about detecting if tr is a patch or not. We should do this only if tr is not a patch.  Let's add a TODO

// TODO:  add ways to detect cleanly if tr is a patch, skip the sync code if it is

File: overlord/configstate/configcore/certs.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+func validateCertSettings(tr config.Conf) error {
+	for _, name := range tr.Changes() {
+		if !strings.HasPrefix(name, "core.certs.") {
+			continue
+		}
+
+		nameWithoutSnap := strings.SplitN(name, ".", 2)[1]
+		cert, err := coreCfg(tr, nameWithoutSnap)
```
Comment (pedronis) [inline overlord/configstate/configcore/certs.go]: we probably need to put some constraints on name here. The config code checks only the provided top level names.

File: overlord/configstate/configcore/certs.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+func validateCertSettings(tr config.Conf) error {
+	for _, name := range tr.Changes() {
+		if !strings.HasPrefix(name, "core.certs.") {
```
[the remark this answers is not in the document]
Comment (mvo5) [reply overlord/configstate/configcore/certs.go]: Yes, my bad, renamed.

File: overlord/configstate/configcore/certs.go (hunk 2 above)
Comment (mvo5) [reply overlord/configstate/configcore/certs.go]: Indeed, added that now

File: tests/main/snapd-certs/task.yaml [hunk 7]
```diff
@@ -0,0 +1,35 @@
+summary: Ensure that store ssl certs for snapd work
+
+systems: [ubuntu-1*, ubuntu-2*]
+
+environment:
+    CERT: /etc/ssl/certs/ssl-cert-snakeoil.pem
+
+execute: |
+    echo "Ensure adding a cert1 to the config works"
```
[the remark this answers is not in the document]
Comment (mvo5) [reply tests/main/snapd-certs/task.yaml:19]: This is added now

File: overlord/configstate/configcore/certs.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+				block, pemCerts = pem.Decode(pemCerts)
+				if block == nil {
+					break
+				}
+				if block.Type != "CERTIFICATE" || len(block.Headers) != 0 {
+					continue
+				}
+				_, err := x509.ParseCertificate(block.Bytes)
+				if err != nil {
+					return fmt.Errorf("cannot parse pem certificate %q", optionName)
```
Comment (pedronis) [inline overlord/configstate/configcore/certs.go:124]: maybe we should have a test that triggers this one

Comment (mvo5) [reply overlord/configstate/configcore/certs.go:124]: Will look into this, *maybe* in a followup but I will see what I can do.

Comment (mvo5) [reply overlord/configstate/configcore/certs.go]: Test added

**Pull request merged**